### ISSUE TITLE
capture stack trace for errors

### ISSUE THREAD

**Issue by mxriverlynn** [oracle/node-oracledb#465]
I'm working on a multi-layered, transactional save for an object graph with about three levels of objects in it.

As I'm building this out, I get errors because I have the wrong data type in my bindvars for the stored procs, or because I misspelled a var in the bindvars, etc.

When this happens, I get an error message, such as:

```
Error: NJS-011: encountered bind value and type mismatch in parameter 2
    at Error (native)
POST /api/entries/save 500 192.169 ms - 107
```

While this error message does tell me that I have a problem with my parameter binding, and it does tell me which parameter, it doesn't tell me which of the many saves in my code is having the problem.

When this happens, I have to debug into my code and hope I put breakpoints in the right places... or use console.log statements in my `if (err){ ... }` code to try and figure out which stored proc call is failing.

It takes a while to find the problem and fix it, because of this. And when I have a second error in my bindings, but it produces the same error message, I don't know if I solved the original one or not because I can't see where the problem originates.

It's very frustrating and time consuming.

fortunately, there's a fix for this - one that should be integrated into the library: `Error.captureStackTrace`

By adjusting my own code, I can work around the problem of not knowing where the problem occurred.

For example:

```
    connection.execute(..., function(err, result){
      if (err) { 
        Error.captureStackTrace(err);
        // ... do something with the error 
      }
  });
```

now, when my error handling code captures this, and I look at `err.stack`, I see something more like this:

```
Error: NJS-011: encountered bind value and type mismatch in parameter 2
    at /Users/derickbailey/dev/glmetadata/lib/glentries/entry/persistence.js:108:15
    at /Users/derickbailey/dev/glmetadata/node_modules/ocarina/lib/storedProc.js:30:23
    at custExecuteCb (/Users/derickbailey/dev/glmetadata/node_modules/oracledb/lib/connection.js:98:7)
```

line 108 of the persistence.js file is where I called `Error.captureStackTrace`. And while this in itself isn't ideal, it gives me the visibility I need to see where the error is, in my code. It greatly reduces the amount of time needed to find the problem, fix it, and know that I fixed this problem when another error pops up.

**Please add the use of `Error.captureStackTrace` to the core library, anywhere an error is created and returned.**

It will be a huge improvement in debugging and working with this library.

**Comment by cjbj** [comment]
Good idea but is `Error.captureStackTrace(err)` the solution for the add-on?  If it is used in node-oracledb, it doesn't report the application script line.

For example, if I modify my application code to capture the stack, the error is useful:

```
$ node issue465-bindfail.js 
Error: ORA-01008: not all variables bound

    at executeCbFunc (/Users/cjones/n/issue465-bindfail.js:21:11)
    at custExecuteCb (/Users/cjones/n/node_modules/oracledb/lib/connection.js:98:7)
```

But if I capture the stack in the error condition in `lib/connection.js` at about line 99 then I only get:

```
$ node issue465-bindfail.js 
Error: ORA-01008: not all variables bound

    at custExecuteCb (/Users/cjones/n/node_modules/oracledb/lib/connection.js:98:13)
```

**Comment by dmcghan** [comment]
I spent a few cycles on this today. The problem has to do with how the stack is lost once an async call is made. Calling `Error.captureStackTrace` is only getting us back to the point the async callback function was invoked - which started a new stack.

Node.js exhibits the same behavior. Here's an example where I try to read a file that doesn't exist using `fs.readFileSync`:
``` javascript
var fs = require('fs');
var file;

try {
  file = fs.readFileSync('doesnotexist.txt', 'utf8');
} catch (e) {
  console.log(e.stack);
}
```

Here's what  I get when I run it:
```bash
Daniels-iMac:~ danielmcghan$ node main.js 
Error: ENOENT: no such file or directory, open 'doesnotexist.txt'
    at Error (native)
    at Object.fs.openSync (fs.js:549:18)
    at Object.fs.readFileSync (fs.js:397:15)
    at Object.<anonymous> (/Users/danielmcghan/main.js:5:13)
    at Module._compile (module.js:410:26)
    at Object.Module._extensions..js (module.js:417:10)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Function.Module.runMain (module.js:442:10)
    at startup (node.js:136:18)
```

As you can see, the stack from a synchronous call is very helpful.

But if we swap out the sync call for an async call:
```javascript
var fs = require('fs');

fs.readFile('doesnotexist.txt', function(err, ignore) {
  console.log(err.stack);
});
```

Then the stack looks a lot like what you reported:
```bash
Daniels-iMac:~ danielmcghan$ node main.js 
Error: ENOENT: no such file or directory, open 'doesnotexist.txt'
    at Error (native)
```

The problem in this case is that I don't know where the problem occurred in "my" code. That's because an error hasn't occurred in my code, I've essentially masked it.

As you mentioned, one way to get the current stack added to an object is with `Error.captureStackTrace`. But as Chris pointed out, that's less useful if we do it as the stack hasn't made it back to your custom code yet.

Another way to get the info you're looking for is to re-throw the error:
```javascript
var fs = require('fs');

fs.readFile('doesnotexist.txt', function(err, ignore) {
  if (err) {
    throw err;
  }
});
```

When I run this I get:
```bash
Daniels-iMac:~ danielmcghan$ node main.js 
/Users/danielmcghan/main.js:5
    throw err;
    ^
 
Error: ENOENT: no such file or directory, open 'doesnotexist.txt'
    at Error (native)
```

Of course what you do with errors passed via callbacks would depend on what you're coding. For example, if using express, you wouldn't throw the error, it would be passed along to `next` to invoke the error handler.

If throwing is not an option, perhaps an error logging module that captures the stack trace for you would be best? At least then you wouldn't have to keep writing `Error.captureStackTrace`.

**Comment by sosoba** [comment]
1. You must create stack before first "low-level" async call and fill given error with them. Draft:
```js
const stack = new Error().stack;
try {
  await conn.execute(sql, params);
} catch (e) {
   e.stack = stack;
   e.sql = sql;
   e.params = params;
   throw e;
}
```

2. I suggest to add more usefull debug data to `Error` object throwed by `oracledb`. In particular `sql` and bind params.

What you think?

**Comment by cjbj** [comment]
Any improvement is good.  Do you want to flesh it out more and show something working?  If so, can you sign the OCA first? https://github.com/oracle/node-oracledb/blob/master/CONTRIBUTING.md

Note we may not be able to include bind parameters "easily".  Last time this was discussed with the Security team they wanted a clear distinction between production and development users, and didn't want binds to be showable in production apps.  They seemed to be OK with binds being accessed if the driver was built with `--debug` - but the discussion was a while ago.

**Comment by markddrake** [comment]
Chris 

Having worked with all the major RDBMS node implementations oracledb seems to be the only one that does not include any kind of stack trace when raising an error (Although as Dan states, with async calls the stack trace is often not that useful, as it terminates in "process ticks and rejections" or similar). However it is better than what is provided by oracledb.

**Comment by cjbj** [comment]
@markddrake play with something like this (with acks to @sosoba's comment above):
```
diff --git a/lib/connection.js b/lib/connection.js
index 92ac44fc..a6db7a57 100644
--- a/lib/connection.js
+++ b/lib/connection.js
@@ -49,7 +49,17 @@ async function execute(sql, a2, a3) {
       break;
   }
 
-  const result = await this._execute(sql, binds, executeOpts);
+  const stack = new Error().stack;
+  let result;
+  try {
+    result = await this._execute(sql, binds, executeOpts);
+  } catch (err) {
+    err = nodbUtil.createErrorWithStack(err, stack);
+    err.sql = sql;
+    if (executeOpts !== undefined)
+      err.executeOpts = executeOpts;
+    throw err;
+  }
 
   // process queries; if a result set is not desired, fetch all of the rows
   // from the result set and then destroy the result set
```
and:
```
diff --git a/lib/util.js b/lib/util.js
index 118554b8..eeee968d 100644
--- a/lib/util.js
+++ b/lib/util.js
@@ -141,6 +141,22 @@ function getErrorMessage(errorCode) {
 
 module.exports.getErrorMessage = getErrorMessage;
 
+// Add a call stack to an Error object.
+// This creates a new Error object since err.message, although accessible, is
+// not shown by vanilla console.error(err) output if "stack" is directly added to 'err'.
+function createErrorWithStack(err, stack) {
+  const e2 = new Error();
+  e2.message = err.message;
+  if (err.errorNum !== undefined)
+    e2.errorNum = err.errorNum;
+  if (err.offset !== undefined)
+    e2.offset = err.offset;
+  e2.stack = stack;
+  return e2;
+}
+
+module.exports.createErrorWithStack = createErrorWithStack;
+
 // assert is typically used at the beginning of public functions to assert
 // preconditions for the function to execute. Most commonly it is used to
 // validate the number of arguments and their types and throw an error if they
```

It seems fugly to penalize every non-erroring `execute()` with the cost of getting the stack.

PR's welcome.

**Comment by cjbj** [comment]
I updated my hack patch comment with a different implementation from the one described by @sosoba.  With Node.js 14+, the stack is shown.

**Comment by cjbj** [comment]
The main branch has the change from @sla100 to capture the stack.  See https://github.com/oracle/node-oracledb/commit/ca165e95b8c9e5ddbaf428550435b361d7b0a0a5  This will be released in the future node-oracledb 5.4.
